. (#3)
morteako/monoids-functional.christmas: monoids.md

This article will introduce an important and interesting concept in functional programming : Monoids.

# What are monoids?

*Monoid* is a concept that comes from abstract algebra, and is there defined as :
a set equipped with an associative binary operation and an identity element.

> A binary operation on a set is a function takes two arguments from the set and produces another one.

Reviewer: ```suggestion
A binary operation on a set is a function that takes two arguments from the set and produces another one.
```

> Being associative means that the grouping of parentheses does not matter, so for a binary operation <> that means that `x <> (y <> z) (x <> y) <> z`.

Reviewer: Leste dette som ett expression først, som var litt forrvirrende, men antar at du mener at disse er like.

```suggestion
Being associative means that the grouping of parentheses does not matter, so for a binary operation <> that means that `x <> (y <> z)` is the same as `(x <> y) <> z`.
```

> Product {getProduct = 30}

foldMap (foldMap (const (Sum 1))) ["abc","df"]
> Sum {getSum = 5}
```

`Maybe` is also a foldable container.

```haskell
instance Foldable Maybe where
    foldMap _ Nothing = mempty
    foldMap f (Just x) = f x

foldMap reverse Nothing :: String
> ""

foldMap Sum (Just 5)
> Sum 5

foldMap (foldMap Product) [Just 2, Nothing, Just 4]
> Product  8
```

When we dont't need to transform the values inside the `Foldable` , we can use the `fold` function.

```haskell
fold :: (Foldable t , Monoid m) => t m -> m
fold = foldMap id

fold (Just "abc")
> "abc"

fold Nothing
> ""

fold [Sum 1, Sum 3, Sum 5]
> Sum 9
```

Now we have seen how to fold data structures using monoids, so let us take a look at some more monoid instances.

### Any - All
 
All and Any are `Bool` newtypes for monoid instances for (&&) with mempty = True and (||) with mempty = False, respectivly.

They can be use to check that any or all elements in a foldable container satisfies a predicate.

```haskell
foldMap (Any . (==5)) [1,3,5,2]
> Any True

foldMap (All . (>2) . length) ["abc","xs"]
> All False
```

Bool can also be made into a monoid using the `xor`-operation.

## Min Max

> When we want to find the maximum of values of values, we can use the `Max` monoid.

Reviewer: Antar at det ikke skal være to "of values".

```suggestion
When we want to find the maximum of values, we can use the `Max` monoid.
```

instance Ord a => Semigroup (Min a) where
  Min a <> Min b = Min (min a b)
  
instance (Ord a, Bounded a) => Monoid (Min a) where
  mempty = maxBound

mempty <> Min 2 <> Min 3
> Min 2
```

As we can see, the identity element for `Min` is `maxBound` and `Max` is `minBound`.

```haskell
mempty :: Min Int
> Min {getMin = 9223372036854775807}
mempty :: Max Int
> Max {getMax = -9223372036854775808}
```

This means that if we fold an empty container, we get a value which is confusing in some cases.

```haskell
fold [] :: Min Int
> Min {getMin = 9223372036854775807}
```

Instead we sometimes want to represent this "failure" using Maybe.

## Lifting a Semigroup into a Monoid

If we take a look at the Maybe data type : `data Maybe a = Nothing | Just a`.

We can see that it can represent a set of values (`Just a`) with an additional element `Nothing`.

If the `a`'s are monoids, we can combine them using `<>` and then chose `Nothing` as our identity element.
In other words, we embeded a semigroup in a monoid by adjoining an identity.

The monoid instance for `Maybe` implements this idea:

```haskell

instance Semigroup a => Semigroup (Maybe a) where
    Nothing <> b       = b

>     a       <> Nothing = a

Reviewer: Her lurer jeg på hvorfor vi må definere begge disse, hvis det var slik at `(<>)` er assosiativ, så burde det vel ikke ha noe å si hvilken vei disse er? Er det verd å snakke om dette? Eller misforstår jeg?

    Just a  <> Just b  = Just (a <> b)

instance Semigroup a => Monoid (Maybe a) where
    mempty = Nothing

fold [Just (Sum 1), Just (Sum 2),Nothing]
> Just (Sum 3)
```

We can use this to fold into a semigroup.
An example of a semigroup that is not a monoid is `First`, which has the semigroup implementation `a <> _ = a`, meaning that it will always pick the left argument.
There is understandably no identity element for this semigroup,
but by using `Maybe (First Int)` we use `Nothing` as our identity.
We have now created a monoid from the First semigroup, and can therefore use it in a fold.

```haskell
foldMap (Just . First) [1,2,3]
> Just (First 1)
```

Back to our "problem" with the min/max identity elements:

> If we now wrap up our type inside a `Maybe`, the result is more sensible(?)

Reviewer: ```suggestion
If we now wrap up our type inside a `Maybe`, the result is more sensible.
```

Ja, det synes jeg.

```haskell
>fold [] :: Maybe (Min Int)
Nothing
```

## Lists

Lists are monoids with `(<>) = (++)` and `mempty = []`.

```haskell
[1,2,3] <> [4,5]
> [1,2,3,4,5]

foldMap (\x -> [x,x*2]) [1,2,3]
> [1,2,2,4,3,6]

```

In fact, the list is an important monoid, called the free monoid.
This comes from the fact that if you take the requirements for a monoid and create the minimal strcuture needed to satisfy these requirements, 
you end up with list monoid instance.

### Tuples

> A very nice neat property about monoids is that they compose over products.

Reviewer: Er det overkill med både nice og neat? Bedre med bare neat kanskje?

Author: haha. ikke med vilje

> Which in practice means, that if you have two monoids, you can put combine them in a tuple and then that tuple is also a monoid.

Reviewer: ```suggestion
Which in practice means, that if you have two monoids, you can combine them in a tuple, and then that tuple is also a monoid.
```

Reviewer: Grammarly mente at det burde være en komma her :)

```haskell
instance (Monoid a, Monoid b) => Monoid (a,b) where
    (a,b) <> (a',b') = (a<>a',b<>b')
    
instance (Monoid a, Monoid b) => Monoid (a,b) where
        mempty = (mempty, mempty)

("abc",Sum 2) <> ("abc",Sum 1)
> ("abcabc", Sum 3)
```

> (This also implemented for tuples of size 3,4 and 5)

Reviewer: ```suggestion
(This is also implemented for tuples of size 3,4 and 5)
```

This makes it easy to accumulate multiple monoid values in a single fold.

> We can use this to find the minimum,maximum,length,sum and product of a list, all in one pass in a neat way. 

Reviewer: Fint med litt mellomrom. Grammarly ville ha en komma før and her også. Jeg stoler på grammarly 😄 

```suggestion
We can use this to find the minimum, maximum, length, sum, and product of a list, all in one pass in a neat way. 
```

Reviewer: Hmm, er lov med begge.

https://www.grammarly.com/blog/comma-before-and/#:~:text=Whether%20or%20not%20you%20put,before%20and%20in%20a%20list.

```haskell
foldMap (\x -> (Min x, Max x, Sum 1, Sum x, Product x)) [5,1,3]
> (Min 1,Max 5,Sum 3,Sum 9,Product 15)
```

And for another example:

> Let us say that we have a list of x y coordinates and want to find AVGRENNSENDE OMRÅDET, the bottom left and top right corner.

Reviewer: ```suggestion
Let us say that we have a list of x y coordinates and want to find the bounding box(?), the bottom left and top right corner.
```

namely min x, min y and max x, max y

```haskell

> > foldMap (\(x,y) -> ((Min x, Min y),(Max x, Max y))) [(0,5),(5,0),(5,10),(7::Int,3::Int)]

Reviewer: Dette er kult!

> Ap {getAp = ZipList [Sum 11,Sum 13]}

```

## Set and Map
Both sets (Data.Set) and maps (Data.Map) have monoid instances, with
their respective `union` as the binary operator and `empty` as the identity element.

```haskell
foldMap Set.fromList [[1,2,3],[3,4,5],[4,5,6]]
> Set.fromList [1,2,3,4,5,6]

Map.fromList [(0,"abc"),(1,"xyz")] <> Map.fromList [(1,"www"),(2,"q")]
> Map.fromList [(0,"abc"),(1,"xyz"),(2,"q")]
```

As we can see, the Map monoid instance only keeps the values from the left map when encountering dupliacate keys.
Often we would like to combine the values instead.
One alternative in this case is to use `Map.unionsWith` instead of `<>`

```haskell
Map.unionsWith (<>) Map.fromList [(0,"abc"),(1,"xyz")] Map.fromList [(1,"www"),(2,"q")]
> Map.fromList [(0,"abc"),(1,"xyzwww"),(2,"q")]
```

But sometimes we actually need this to be a monoid instance.
The [appendmap](https://hackage.haskell.org/package/appendmap) package provides a newtype `AppendMap` for exactly this purpose.

```haskell
AppendMap (Map.fromList [(0,"abc"),(1,"xyz")]) <>  AppendMap (Map.fromList [(1,"www"),(2,"q")])
> AppendMap Map.fromList [(0,"abc"),(1,"xyzwww"),(2,"q")]
```

# Summary

> We have now looked at different Semigroups and Monoids how they are useful when working with Foldable containers.

Reviewer: ```suggestion
We have now looked at different Semigroups and Monoids, and how they are useful when working with Foldable containers.
```

Reviewer: Usikker på om det burde være komma eller ikke her

Reviewer: Men and burde det nok være

Hopefully you learned something new or something became clearer than it was before.
```haskell
foldMap fold [Just "Merry", Nothing, Just " ", Nothing ,Just "Christmas"] <> "!"

> > "Merry Christmas!"`

Reviewer: Morsom avslutning!

left and top right corner, i.e. (min x, min y) and (max x, max y).

```haskell
> foldMap (\(x,y) -> ((Min x, Min y),(Max x, Max y))) [(0,5),(5,0),(5,10),(7::Int,3::Int)]
((Min 0,Min 0), (Max 7,Max 1))
```

### Ap -  Lifting a Monoid inta an Applicative

Applacatives have a strong connection to monoids. In categoory theory they are called `lax monoidal functors`.
Every Applicative gives us a monoid, by defining
`(<>) = liftA2 (<>)` and `mempty = pure mempty`.
This is the instance for the `Ap` newtype.

```haskell
[Sum 1,Sum 2] <> [Sum 10, Sum 11,Sum 12]
> [Sum 1,Sum 2,Sum 10,Sum 11,Sum 12]

Ap [Sum 1,Sum 2] <> Ap [Sum 10, Sum 11,Sum 12]
> Ap {getAp = [Sum 11,Sum 12,Sum 13,Sum 12,Sum 13,Sum 14]}
```

We see here how `<>` for regular lists, just appends the lists toghether, but 
`<>` for `Ap []` combines every element in the first list with every element in the second list,
which mirrors the applicative instance for lists.  

```haskell
> fold [Just "a", Nothing, Just "c"]
Just "ac"
> fold [Ap $ Just "a", Ap $ Nothing, Ap $ Just "c"]
Ap {getAp = Nothing}
```

As we have seen earlier, in the regular Maybe monoid, `Nothing` is the identity element.

> Moreover, the applicative instance for maybe short circuits when you encounter `Nothing

Reviewer: ```suggestion
Moreover, the applicative instance for maybe short circuits when you encounter `Nothing`
```

Reviewer: Her manglet det en `, som gjorde at markup-en ble feil.
